File: Array/huangxinyu/c/MaxQueue.cpp
//
// Created by hasee on 2020/6/19.
//
#include <queue>

class MaxQueue {
public:
    MaxQueue() {

    }

    int max_value() {
        if (queue.empty()) {
            return -1;
        }
        return deque.front();
    }

    void push_back(int value) {
        queue.push(value);
        while (deque.back() <= value) {
            deque.pop_back();
        }
        deque.push_back(value);
    }

    int pop_front() {
        if (queue.empty()) {
            return -1;
        }
        int i = queue.front();
        queue.pop();
        if (i == deque.front()) {
            deque.pop_front();
        }
        return i;
    }

private:
    std::queue<int> queue;
    std::deque<int> deque;
};

File: Array/huangxinyu/kotlin/src/MaxQueue.java
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.LinkedList;
import java.util.Queue;

public class MaxQueue {
    private Deque<Integer> queue;
    private Deque<Integer> help;

    public MaxQueue() {
        queue = new ArrayDeque<>();
        help = new ArrayDeque<>();
    }

    public int max_value() {
        return queue.isEmpty() ? -1 : help.peek();
    }

    public void push_back(int value) {
        queue.offer(value);
        while(!help.isEmpty() && value > help.peekLast()) {
            help.pollLast();
        }
        help.offer(value);
    }

    public int pop_front() {
        if(queue.isEmpty()) {
            return -1;
        }
        int val = queue.pop();
        if(help.peek() == val) {
            help.pop();
        }
        return val;
    }
}
